### PR TITLE
drivers: sensor: hc-sr04: add driver

### PR DESCRIPTION
**Pull request by Pixep**
Add a simple driver for the HC-SR04 ultrasonic distance sensor. This sensor only uses two GPIO pins (trigger and echo) to trigger the acquisition. The signal received on the `echo` pin is the duration between the emission of the ultrasonic pulse and the feedback received (within range of the sensor).

### CODE REVIEW

File: drivers/sensor/hc_sr04/hc_sr04.c [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+		.dev = DEVICE_DT_INST_GET(index),                                     \
+		.start_cycles = 0,                                                    \
+		.echo_high_cycles = ATOMIC_INIT(0),                                   \
+	};                                                                            \
+	static struct hcsr04_config hcsr04_config_##index = {                         \
+		.trigger_gpios = GPIO_DT_SPEC_INST_GET(index, trigger_gpios),         \
+		.echo_gpios = GPIO_DT_SPEC_INST_GET(index, echo_gpios),               \
+	};                                                                            \
+                                                                                      \
+	DEVICE_DT_INST_DEFINE(index, &hcsr04_init, NULL, &hcsr04_data_##index,        \
```
Comment (MaureenHelm) [inline drivers/sensor/hc_sr04/hc_sr04.c:170]: Use `SENSOR_DEVICE_DT_INST_DEFINE`

File: drivers/sensor/hc_sr04/hc_sr04.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	struct gpio_callback gpio_cb;
+	struct k_sem sem;
+	uint32_t start_cycles;
+	atomic_t echo_high_cycles;
+};
+
+struct hcsr04_config {
+	struct gpio_dt_spec trigger_gpios;
+	struct gpio_dt_spec echo_gpios;
+};
```
Comment (MaureenHelm) [inline drivers/sensor/hc_sr04/hc_sr04.h:26]: There's not enough here to justify having a separate private include header, please move these to the .c file.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Add this sensor to `tests/drivers/build_all/sensor`

File: drivers/sensor/hc_sr04/hc_sr04.c (hunk 1 above)
Comment (Pixep) [reply drivers/sensor/hc_sr04/hc_sr04.c]: Didn't know about it! Thanks, done.

File: drivers/sensor/hc_sr04/hc_sr04.h (hunk 2 above)
Comment (Pixep) [reply drivers/sensor/hc_sr04/hc_sr04.h]: Done.

**Comment by Pixep** [comment]
> Add this sensor to `tests/drivers/build_all/sensor`

Done. Let me know if you see any improvements to make!

**Comment by Pixep** [comment]
Hi, this is still active, and the requested changes have been made.
Happy to rebase if needed.

**Comment by Pixep** [comment]
Hi @MaureenHelm , when you have time, would you be able to take a look and let me know if the update is going in the right direction?
I would be happy to rebase and update it further, as long as this driver is of interest to Zephyr.

Thank you!

File: tests/drivers/sensor/hc-sr04/prj.conf [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+CONFIG_SENSOR_LOG_LEVEL_WRN=y
+
+# Enable sensors
+CONFIG_SENSOR=y
+
+# Enable GPIO
+CONFIG_GPIO=y
+
+# Enable HC-SR04 sensor
+CONFIG_HC_SR04=y
```
Comment (ubieda) [inline tests/drivers/sensor/hc-sr04/prj.conf:16]: Should be enabled by default with `CONFIG_SENSOR` and having it defined in device-tree
```suggestion
```

File: tests/drivers/sensor/hc-sr04/boards/native_sim.overlay
Comment (ubieda) [inline tests/drivers/sensor/hc-sr04/boards/native_sim.overlay:1]: > > Add this sensor to tests/drivers/build_all/sensor
> 
> Done. Let me know if you see any improvements to make!

I don't see it added yet. Please revisit this.

**Comment by Pixep** [comment]
Thank you @ubieda, I missed the distinction between build and functional tests and added the build one.

File: tests/drivers/sensor/hc-sr04/prj.conf (hunk 3 above)
Comment (Pixep) [reply tests/drivers/sensor/hc-sr04/prj.conf]: Done

File: tests/drivers/sensor/hc-sr04/boards/native_sim.overlay
Comment (Pixep) [reply tests/drivers/sensor/hc-sr04/boards/native_sim.overlay:1]: Thank you Luis, I missed the distinction. Done.

**Review by MaureenHelm — APPROVED** [review]

**Review by ubieda — APPROVED** [review]

**Pull request merged**